### PR TITLE
build: Update maven plugin versions

### PR DESCRIPTION
**Pull request by rleigh-codelibre**
Update all maven plugins to the latest version:

- Update all maven modules (newer versions have support for Java 9; the old versions we use are broken)
- Bump minimum maven version (required for the new plugin versions)
- Travis and appveyor build with JDK9 as well as 7 and 8.

Testing: Check travis builds are all green.  Check appveyor builds are all green (added in this PR, tested on my personal branches; needs adding to the ome repositories as a followup).  Check plugin updates are sane with e.g. `mvn versions:display-plugin-updates`

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -6,10 +6,18 @@ sudo: false
...
   - openjdk7
 
 matrix:
   fast_finish: true
+
+before_install:
+ - mkdir -p downloads
+ - if [[ ! -f downloads/maven.tar.gz ]]; then curl -o downloads/maven.tar.gz http://mirror.ox.ac.uk/sites/rsync.apache.org/maven/maven-3/3.5.0/binaries/apache-maven-3.5.0-bin.tar.gz; fi
+ - tar xf downloads/maven.tar.gz
+ - export PATH=$(pwd)/apache-maven-3.5.0/bin:$PATH
```
Comment (sbesson) [inline .travis.yml]: From https://travis-ci.org/ome/ome-common-java/jobs/279470866

```
mvn version

Apache Maven 3.5.0 (ff8f5e7444045639af65f6095c62210b5713f426; 2017-04-03T19:39:06Z)
Maven home: /usr/local/maven-3.5.0
```

so the latest Travis images ship with Maven 3.5.0 by default. What is the rationale behind this `before_install`?

Comment (rleigh-codelibre) [reply .travis.yml]: This work dates from June, and the maven provided by travis was an old version which didn't work with java 9 at all.  If the images are updated, we can drop this entirely.

**Comment by mtbc** [comment]
The Travis logs for this and the related PRs do include, "Switching to Oracle JDK9 (java-9-oracle)", "Java version: 9, vendor: Oracle Corporation" and other such encouragement.

**Comment by mtbc** [comment]
https://github.com/ome/ome-model/pull/53 has a substantial version bump for `build-helper-maven-plugin`.

File: pom.xml [hunk 2]
```diff
@@ -72,7 +72,7 @@
   </mailingLists>
 
   <prerequisites>
-    <maven>3.0</maven>
+    <maven>3.0.5</maven>
```
Comment (mtbc) [inline pom.xml]: Debian 7's back on maven 3.0.4 but that's pretty ancient so if we do need 3.0.5 then go for it.

**Comment by mtbc** [comment]
I concur with @sbesson about checking release notes at least for the highlights but I am very open to being persuaded that our CI coverage is good enough to pick out problems with these updates.

(I hope it's okay if I don't comment separately on the related PRs, it would largely just be copy and paste.)

**Comment by rleigh-codelibre** [comment]
Version changes across all repos:

```
build-helper-maven-plugin           1.12 → 3.0.0
exec-maven-plugin                  1.5.0 → 1.6.0
license-maven-plugin            1.9,1.13 → 1.14
maven-assembly-plugin          2.6,3.0.0 → 3.1.0
maven-compiler-plugin       3.5.1, 3.6.1 → 3.7.0
maven-dependency-plugin             2.10 → 3.0.2
maven-gpg-plugin                     1.5 → 1.6
maven-javadoc-plugin       2.9.1, 2.10.4 → 3.0.0-M1
maven-resources-plugin             3.0.1 → 3.0.2
maven-site-plugin                  3.5.1 → 3.6
maven-surefire-plugin             2.19.1 → 2.20.1
nexus-staging-maven-plugin         1.6.7 → 1.6.8
versions-maven-plugin                2.3 → 2.4
```

I'll follow up with what the changes are.

**Comment by rleigh-codelibre** [comment]
All plugin updates include a mixture of minor bugfixes and updates to their own dependencies and maven plugins.

All major version bumps (2 to 3) are due to the dropping of Maven 2 support, making Maven 3 the minimum versions.  This is often accompanied by the dropping of Java 6 support.

## build-helper-maven-plugin
https://github.com/mojohaus/build-helper-maven-plugin/compare/build-helper-maven-plugin-1.12...build-helper-maven-plugin-3.0.0

- Upstream moved from 1.12 directly to 3.0.0
- Drops maven 2 support
- Drops java 6 support

## exec-maven-plugin
https://github.com/mojohaus/exec-maven-plugin/compare/exec-maven-plugin-1.5.0...exec-maven-plugin-1.6.0

- Drops java 6 support
- Adds java 9 support (jigsaw module-info)

## license-maven-plugin
https://github.com/mojohaus/license-maven-plugin/compare/license-maven-plugin-1.9...license-maven-plugin-1.14

- A lot of bug fixes

## maven-assembly-plugin
https://issues.apache.org/jira/projects/MASSEMBLY/versions/12330359
https://issues.apache.org/jira/projects/MASSEMBLY/versions/12338667

- Drops java 6 support
- Drops maven 2 support

## maven-compiler-plugin
https://issues.apache.org/jira/projects/MCOMPILER/versions/12334708
https://issues.apache.org/jira/projects/MCOMPILER/versions/12338627
https://issues.apache.org/jira/projects/MCOMPILER/versions/12339154
https://issues.apache.org/jira/projects/MCOMPILER/versions/12341266

- Java 9 support (compile, module-info)

## maven-dependency-plugin
https://issues.apache.org/jira/projects/MDEP/versions/12330458
https://issues.apache.org/jira/projects/MDEP/versions/12338874
https://issues.apache.org/jira/projects/MDEP/versions/12341254

- Drops maven 2 support
- Java 9 support

## maven-gpg-plugin
https://issues.apache.org/jira/projects/MGPG/versions/12330780

- Drops maven 2 support
- Avoids need for passphrases for noninteractive use

## maven-javadoc-plugin
https://issues.apache.org/jira/projects/MJAVADOC/versions/12330871
https://issues.apache.org/jira/projects/MJAVADOC/versions/12330872
https://issues.apache.org/jira/projects/MJAVADOC/versions/12330872
https://issues.apache.org/jira/projects/MJAVADOC/versions/12330876
https://issues.apache.org/jira/projects/MJAVADOC/versions/12331967

- Java 8 support
- Java 9 support (in 2.10.4)
- further Java 8 support (in 3.0.0 -- should allow removal of current special casing)

The 3.0.0-M1 is likely unnecessary for Java 9 support, but would allow
for the removal of nasty use of profiles to build Javadocs with Java
8+.

## maven-resources-plugin
https://issues.apache.org/jira/projects/MRESOURCES/versions/12336059

- Minor bugfixes and plugin updates

## maven-site-plugin
https://issues.apache.org/jira/projects/MSITE/versions/12335570

- Minor bugfixes and plugin updates

## maven-surefire-plugin
https://issues.apache.org/jira/projects/SUREFIRE/versions/12334636
https://issues.apache.org/jira/projects/SUREFIRE/versions/12340297

- Drops Java 6 support
- Java 9 support

## nexus-staging-maven-plugin

- Minor bugfixes and plugin updates

## versions-maven-plugin
https://github.com/mojohaus/versions-maven-plugin/compare/versions-maven-plugin-2.3...versions-maven-plugin-2.4

- Minor bugfixes and plugin updates

**Comment by rleigh-codelibre** [comment]
Testing the javadoc plugin with 3.0.0-M1 and 2.10.4, both work with Java 9 and both also work with Java 9 with the special case Java 8+ profile removed.  But only for OpenJDK.  3.0.0-M1 is required for Oracle Java 9.

**Comment by sbesson** [comment]
Thanks for providing https://github.com/ome/ome-common-java/pull/16#issuecomment-332859611. I will look at the release notes as soon as I get some time. A current unknown as discussed last Wednesday are more about the timeline for the `maven-javadoc-plugin` 3.0.0 release.

In terms of stack integration, this set of PR  distributed across low-level Java components echo the interrogations raised by @melissalinkert in https://github.com/ome/ome-codecs/pull/8#issue-258943807 i.e. how to integrate PRs opened against decoupled repositories and test them down the stack before we cut releases. As an initial point of discussion, a couple of suggestions ranging from the simplest to the more complex:

- provide a single `Makefile` or equivalent allowing to create a reproducible build from all the SNAPSHOT and perform some manual testing
- discussed with @melissalinkert, a more systematic approach amending the existing infrastructure could be to script this logic into https://ci.openmicroscopy.org/job/BIOFORMATS-DEV-merge-repository-subset/ and have job additionally merge and build low-level Maven projects before running the automated tests against a subset of the data repository
- a more involved but thorough solution might be to create an ephemerous Continuous Integration server based on the existing [devspace](https://github.com/openmicroscopy/devspace) including Maven jobs for the low-level components and propagating and testing the changes up to the OMERO level.

Other thoughts welcome on how to review our process to facilitate this integration especially if we want this effort to be coordinated with the upcoming work on Java 9.

**Comment by rleigh-codelibre** [comment]
@sbesson Here's a testing script.  Note that while it produces working bioformats jars, the unit tests fail due to `ij.jar` needing an update (waiting on a new release).

```
#!/bin/sh

set -e
set -x

build=/tmp/java9-build

mkdir -p "$build"

for repository in ome-common-java ome-model ome-poi ome-mdbtools ome-jai ome-codecs \
                  ome-metakit ome-stubs bioformats
do
    cd "$build"

    user=ome
    if [ "$repository" = "bioformats" ]; then
        user=openmicroscopy
    fi

    if [ ! -d "$repository" ]; then
        git clone "https://github.com/${user}/${repository}.git" "$repository"
        cd "$repository"
        git remote add rleigh "https://github.com/rleigh-codelibre/${repository}.git"
        git fetch rleigh
    fi

    cd "${build}/${repository}"

    git fetch --all
    git reset --hard
    git clean -dxf
    git checkout rleigh/java9-build

    # Patch to use explicit snapshot versions for all dependencies
    case "$repository" in
        ome-metakit)
            sed -i -e "s;<ome-common\.version>.*</ome-common\.version>;<ome-common.version>5.3.3-SNAPSHOT</ome-common.version>;" pom.xml
            ;;
        ome-model)
            sed -i -e "s;<ome_common\.version>.*</ome_common\.version>;<ome_common.version>5.3.3-SNAPSHOT</ome_common.version>;" pom.xml
            ;;
        ome-poi)
            sed -i -e "s;<version>5\.3\.1</version>;<version>5.3.3-SNAPSHOT</version>;" pom.xml
            ;;
        ome-codecs)
            sed -i -e "s;<version>5\.3\.2</version>;<version>5.3.3-SNAPSHOT</version>;" \
                -e "s;<version>0\.1\.0</version>;<version>0.1.1-SNAPSHOT</version>;" pom.xml
            ;;
        bioformats)
            sed -i -e "s;<ome-common\.version>.*</ome-common\.version>;<ome-common.version>5.3.3-SNAPSHOT</ome-common.version>;" \
                -e "s;<ome-model\.version>.*</ome-model\.version>;<ome-model.version>5.5.8-SNAPSHOT</ome-model.version>;" \
                -e "s;<ome-poi\.version>.*</ome-poi\.version>;<ome-poi.version>5.3.2-SNAPSHOT</ome-poi.version>;" \
                -e "s;<ome-mdbtools\.version>.*</ome-mdbtools\.version>;<ome-mdbtools.version>5.3.1-SNAPSHOT</ome-mdbtools.version>;" \
                -e "s;<ome-jai\.version>.*</ome-jai\.version>;<ome-jai.version>0.1.1-SNAPSHOT</ome-jai.version>;" \
                -e "s;<ome-codecs\.version>.*</ome-codecs\.version>;<ome-codecs.version>0.2.1-SNAPSHOT</ome-codecs.version>;" \
                -e "s;<ome-metakit\.version>.*</ome-metakit\.version>;<ome-metakit.version>5.3.1-SNAPSHOT</ome-metakit.version>;" \
                -e "s;<lwf-stubs\.version>.*</lwf-stubs\.version>;<lwf-stubs.version>5.3.1-SNAPSHOT</lwf-stubs.version>;" \
                -e "s;<mipav-stubs\.version>.*</mipav-stubs\.version>;<mipav-stubs.version>5.3.1-SNAPSHOT</mipav-stubs.version>;" \
                pom.xml
            ;;
    esac

    git diff

    if [ "$repository" = "bioformats" ]; then
        ant clean jars tools
        ant test || true
    fi
    
    mvn install
done
```

I think this script demonstrates the need for a common repository containing the maven dependency versions.  This would provide a single place to set the versions, and then the separate repositories would simply need updating to use a snapshot version of this repository; we would then switch them to use the non-snapshot version when we release them.  Workflow-wise, this would mean when we release we update two versions rather than one, and post-release we set two snapshot versions again (the pom version and the common repo version).

This script runs in just a few minutes, so regarding continuous integration, it would not cost much in terms of build times for us to have this replace the bioformats merge/latest jobs.  The script would be easy to adapt to use scc merge branches instead of my personal branches.  Or it could be reimplemented in terms of jenkins git plugin configuration (the sed patching would be unnecessary once we update our workflow as defined above).

**Comment by rleigh-codelibre** [comment]
This variant of the script isn't for CI use, but for testing differences between the jars.  It builds the original sources with openjdk8 (reference build).  Then it builds the sources with the java9+maven changes with openjdk8, openjdk8 and oraclejdk9.  Then it uses `pkgdiff` to compare the jars for these three variants against the reference build.  These diffs show trivial changes (timestamps, pom changes, manifest.mf changes) as well as changes to the class files.

I'll link to the results once they have finished processing.

```
#!/bin/sh

set -e
set -x

openjdk8="/usr/lib/jvm/java-1.8.0-openjdk-amd64"
openjdk9="/usr/lib/jvm/java-1.9.0-openjdk-amd64"
oraclejdk9="/opt/jdk-9"
savedpath="$PATH"

builddir=/tmp/java9-build

mkdir -p "$builddir"

reference=original-openjdk8
states="$reference new-openjdk8 new-openjdk9 new-oraclejdk9"
repositories="ome-common-java ome-model ome-poi ome-mdbtools ome-jai ome-codecs ome-metakit ome-stubs bioformats"
for state in $states
do
    for repository in $repositories
    do
        echo "Building $repository / $state"

        case "$state" in
            *-openjdk8)
                export JAVA_HOME="$openjdk8"
                ;;
            *-openjdk9)
                export JAVA_HOME="$openjdk9"
                ;;
            *-oraclejdk9)
                export JAVA_HOME="$oraclejdk9"
                ;;
        esac
        PATH="$JAVA_HOME/bin:$savedpath"

        build="${builddir}/${state}"
        mkdir -p "$build"
        cd "$build"

        user="ome"
        branch="master"
        if [ "$repository" = "bioformats" ]; then
            user="openmicroscopy"
            branch="develop"
        fi

        if [ ! -d "$repository" ]; then
            git clone "https://github.com/${user}/${repository}.git" "$repository"
            
            if [ "$state" != "$reference" ]; then
                cd "$repository"
                git remote add rleigh "https://github.com/rleigh-codelibre/${repository}.git"
                git fetch rleigh
            fi
        fi

        cd "${build}/${repository}"

        git fetch --all
        git reset --hard
        git clean -dxf
        if [ "$state" != "$reference" ]; then
            git checkout rleigh/java9-build

            # Patch to use explicit snapshot versions for all dependencies
            case "$repository" in
                ome-metakit)
                    sed -i -e "s;<ome-common\.version>.*</ome-common\.version>;<ome-common.version>5.3.3-SNAPSHOT</ome-common.version>;" pom.xml
                    ;;
                ome-model)
                    sed -i -e "s;<ome_common\.version>.*</ome_common\.version>;<ome_common.version>5.3.3-SNAPSHOT</ome_common.version>;" pom.xml
                    ;;
                ome-poi)
                    sed -i -e "s;<version>5\.3\.1</version>;<version>5.3.3-SNAPSHOT</version>;" pom.xml
                    ;;
                ome-codecs)
                    sed -i -e "s;<version>5\.3\.2</version>;<version>5.3.3-SNAPSHOT</version>;" \
                        -e "s;<version>0\.1\.0</version>;<version>0.1.1-SNAPSHOT</version>;" pom.xml
                    ;;
                bioformats)
                    sed -i -e "s;<ome-common\.version>.*</ome-common\.version>;<ome-common.version>5.3.3-SNAPSHOT</ome-common.version>;" \
                        -e "s;<ome-model\.version>.*</ome-model\.version>;<ome-model.version>5.5.8-SNAPSHOT</ome-model.version>;" \
                        -e "s;<ome-poi\.version>.*</ome-poi\.version>;<ome-poi.version>5.3.2-SNAPSHOT</ome-poi.version>;" \
                        -e "s;<ome-mdbtools\.version>.*</ome-mdbtools\.version>;<ome-mdbtools.version>5.3.1-SNAPSHOT</ome-mdbtools.version>;" \
                        -e "s;<ome-jai\.version>.*</ome-jai\.version>;<ome-jai.version>0.1.1-SNAPSHOT</ome-jai.version>;" \
                        -e "s;<ome-codecs\.version>.*</ome-codecs\.version>;<ome-codecs.version>0.2.1-SNAPSHOT</ome-codecs.version>;" \
                        -e "s;<ome-metakit\.version>.*</ome-metakit\.version>;<ome-metakit.version>5.3.1-SNAPSHOT</ome-metakit.version>;" \
                        -e "s;<lwf-stubs\.version>.*</lwf-stubs\.version>;<lwf-stubs.version>5.3.1-SNAPSHOT</lwf-stubs.version>;" \
                        -e "s;<mipav-stubs\.version>.*</mipav-stubs\.version>;<mipav-stubs.version>5.3.1-SNAPSHOT</mipav-stubs.version>;" \
                        pom.xml
                    ;;
            esac

            git diff
        else
            git checkout "origin/$branch"
        fi

        if [ "$repository" = "bioformats" ]; then
            ant clean jars tools
            ant test || true
        fi
        
        if [ "$repository" = "bioformats" ]; then
            mvn install || true
        else
            mvn install
        fi
    done
done

mkdir -p "$builddir/reports"

for state in $states
do
    if [ "$state" = "$reference" ]; then
        continue
    fi

    mkdir -p "${builddir}/reports/${state}"
    cd "${builddir}/reports/${state}"

    for repository in $repositories
    do
        echo "JARS: " ${builddir}/${state}/${repository}/target/*.jar
        for jarfile in ${builddir}/${state}/${repository}/target/*.jar; do
            basejarfile=$(basename "$jarfile")
            echo "DIFF: $state / $repository / $basejarfile"
            pkgdiff "${builddir}/${reference}/${repository}/target/${basejarfile}" "${builddir}/${state}/${repository}/target/${basejarfile}" || true
        done
    done
done
```

**Comment by rleigh-codelibre** [comment]
Maven javadoc plugin still hasn't had a final release: https://mvnrepository.com/artifact/org.apache.maven.plugins/maven-javadoc-plugin

**Comment by rleigh-codelibre** [comment]
With the release of ImageJ 1.51r, we can now build and test bioformats with Java 9.  I opened https://github.com/openmicroscopy/bioformats/pull/2986 now this last piece of the puzzle is in place.  The whole collection now builds and tests with Java 9.

**Comment by rleigh-codelibre** [comment]
Note the javadoc 3.0.0-M1 version has now been replaced with a proper 3.0.0 release.  I've updated this PR, and I'll also update the other PRs to remove the use of beta plugin versions.

**Comment by rleigh-codelibre** [comment]
Appveyor support is tested here: https://ci.appveyor.com/project/rleigh-codelibre/ome-common-java/build/1.0.2 and can be added to the ome/ome-common-java repository once this PR is merged.

**Comment by rleigh-codelibre** [comment]
I have updated the shell scripts which build the whole set of projects together, and then compare the jar contents for different JDK versions with and without the PRs merged.

**Comment by joshmoore** [comment]
Merging as per https://github.com/ome/ome-codecs/pull/10#issuecomment-370461418 for a patch bump.

Thanks for all the cleanup, @rleigh-codelibre

**Pull request merged**